src/sources/prometheus/remote_write.rs

```diff
@@ -0,0 +1,270 @@
+            emit!(PrometheusRemoteWriteParseError {
+                error: error.clone()
+            });
+            ErrorMessage::new(
+                StatusCode::BAD_REQUEST,
+                format!("Could not decode write request: {}", error),
+            )
+        })?;
+        Ok(decode_request(request))
+    }
+}
+
+impl HttpSource for RemoteWriteSource {
+    fn build_event(
+        &self,
+        body: Bytes,
+        _header_map: HeaderMap,
+        _query_parameters: HashMap<String, String>,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        let byte_size = body.len();
+        let result = self.decode_body(body)?;
+        let count = result.len();
+        emit!(PrometheusRemoteWriteReceived { byte_size, count });
+        Ok(result)
+    }
+}
+
+fn decode_request(request: proto::WriteRequest) -> Vec<Event> {
+    request
+        .timeseries
+        .into_iter()
+        .filter_map(decode_timeseries)
+        .flatten()
+        .collect()
+}
+
+fn decode_timeseries(timeseries: proto::TimeSeries) -> Option<impl Iterator<Item = Event>> {
+    let (name, tags) = parse_labels(timeseries.labels);
+    match name {
+        Some(name) => Some(timeseries.samples.into_iter().map(move |sample| {
+            let value = sample.value;
+            let value = if name.ends_with("_total") {
+                MetricValue::Counter { value }
+            } else {
+                MetricValue::Gauge { value }
+            };
+            Metric {
+                name: name.clone(),
+                namespace: None,
+                timestamp: parse_timestamp(sample.timestamp),
+                tags: tags.clone(),
+                kind: MetricKind::Absolute,
+                value,
+            }
+            .into()
+        })),
+        None => {
+            emit!(PrometheusNoNameError);
+            None
+        }
+    }
+}
+
+fn parse_labels(labels: Vec<proto::Label>) -> (Option<String>, Option<BTreeMap<String, String>>) {
+    let mut tags = labels
+        .into_iter()
+        .map(|label| (label.name, label.value))
+        .collect::<BTreeMap<String, String>>();
+    let name = tags.remove(METRIC_NAME_LABEL);
+    let tags = if tags.is_empty() { None } else { Some(tags) };
+    (name, tags)
+}
+
+fn parse_timestamp(timestamp: i64) -> Option<DateTime<Utc>> {
+    match Utc.timestamp_opt(timestamp / 1000, (timestamp % 1000) as u32 * 1000000) {
+        LocalResult::None => None,
+        LocalResult::Single(timestamp) => Some(timestamp),
+        // If the result is ambiguous, arbitrarily pick one of the possibilities.
+        LocalResult::Ambiguous(timestamp1, _timestamp2) => Some(timestamp1),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn genreate_config() {
+        crate::test_util::test_generate_config::<PrometheusRemoteWriteConfig>();
+    }
+}
+
+#[cfg(all(test, feature = "prometheus-integration-tests"))]
```

Comment:
Does this need `prometheus-integration-tests`? It seems like we are just testing our source with the sink rather than needing an actual prometheus instance to be running.

Reply:
True. I'll move it.